### PR TITLE
[New arch] Stop supporting <10 servers 

### PR DESCRIPTION
**Pull request by davigonz**
Implements #2636 
Needs https://github.com/owncloud/android-library/pull/265

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt [hunk 1]
```diff
@@ -313,10 +313,6 @@ class DocumentsStorageProvider : DocumentsProvider() {
 
     private fun createFolder(parentDocument: OCFile, displayName: String): String {
         val newPath = parentDocument.remotePath + displayName + OCFile.PATH_SEPARATOR
-        val serverWithForbiddenChars = isVersionWithForbiddenCharacters()
-        if (!FileUtils.isValidName(displayName, serverWithForbiddenChars)) {
```
Comment (abelgardep) [inline owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: isValidName is not needed anymore, is it?

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/ShareActivity.kt [hunk 2]
```diff
@@ -217,18 +217,10 @@ class ShareActivity : FileActivity(), ShareFragmentListener {
...
-            } else {
-                if (file.isFolder)
-                    RemoteShare.FEDERATED_PERMISSIONS_FOR_FOLDER_UP_TO_OC9
-                else
-                    RemoteShare.FEDERATED_PERMISSIONS_FOR_FILE_UP_TO_OC9
-            }
+            return if (file.isFolder)
+                RemoteShare.FEDERATED_PERMISSIONS_FOR_FOLDER
+            else
+                RemoteShare.FEDERATED_PERMISSIONS_FOR_FILE
```
[the remark this answers is not in the document]
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/ShareActivity.kt]:
```
private fun getAppropiatePermissions(shareType: ShareType?): Int {
        // check if the Share is FERERATED
        val isFederated = ShareType.FEDERATED == shareType

        return when {
            file.isSharedWithMe -> RemoteShare.READ_PERMISSION_FLAG    // minimum permissions
            isFederated ->
                if (file.isFolder) {
                    RemoteShare.FEDERATED_PERMISSIONS_FOR_FOLDER
                } else {
                    RemoteShare.FEDERATED_PERMISSIONS_FOR_FILE
                }
            else ->
                if (file.isFolder) {
                    RemoteShare.MAXIMUM_PERMISSIONS_FOR_FOLDER
                } else {
                    RemoteShare.MAXIMUM_PERMISSIONS_FOR_FILE
                }
        }
}
```

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/fragments/EditPrivateShareFragment.kt [hunk 3]
```diff
@@ -169,14 +169,9 @@ class EditPrivateShareFragment : DialogFragment() {
             val sharePermissions = share!!.permissions
             val isFederated = ShareType.FEDERATED.value == share!!.shareType
             val serverVersion = AccountUtils.getServerVersion(account)
-            val isNotReshareableFederatedSupported =
-                serverVersion != null && serverVersion.isNotReshareableFederatedSupported
```
[the remark this answers is not in the document]
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/fragments/EditPrivateShareFragment.kt]: It's not needed, removing it...

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/fragments/EditPrivateShareFragment.kt [hunk 4]
```diff
@@ -169,14 +169,9 @@ class EditPrivateShareFragment : DialogFragment() {
...
             var compound: CompoundButton
 
             compound = canShareSwitch
-            if (isFederated && !isNotReshareableFederatedSupported) {
-                compound.visibility = View.INVISIBLE
-            }
```
[the remark this answers is not in the document]
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/fragments/EditPrivateShareFragment.kt]: Sure

File: owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/fragments/EditPrivateShareFragment.kt [hunk 5]
```diff
@@ -274,9 +267,7 @@ class EditPrivateShareFragment : DialogFragment() {
                     if (file?.isFolder == true) {
                         if (isChecked) {
                             val serverVersion = AccountUtils.getServerVersion(account)
-                            val isNotReshareableFederatedSupported =
-                                serverVersion != null && serverVersion.isNotReshareableFederatedSupported
```
[the remark this answers is not in the document]
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/ui/sharing/fragments/EditPrivateShareFragment.kt]: Done

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt (hunk 1 above)
Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Neither `isValidName` nor `isValidPath` of `FileUtils.jave` are needed so removing them

**Comment by davigonz** [comment]
@abelgardep changes applied, could you review also https://github.com/owncloud/android-library/pull/265?

File: owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt (hunk 1 above)
Comment (abelgardep) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Maybe we should keep checking if there are forbiddenCharacters like / * or \

Comment (davigonz) [reply owncloudApp/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.kt]: Sure, done

**Comment by jesmrec** [comment]
To test this feature, i got the edge supported (9.0 -> 9.1) and i compared the behaviour in the main operations there. It would be desirable to check also in 8.x, but i will need to build another container out of dockerhub world.

The results over the main features are these ones:

| Test Case | 9.0.9 | 9.1.8 | 
| :-------- | :---: | :---: | 
| Login http    | OK    |  OK   |
| Login https    | **Crash**    |  **Crash**   |
| Browsing  | OK | OK |  
| Sharing   | **Crash** |  OK   |
| Upload    | Not running (Conflict) | OK | 
| Downloads | OK   |  OK   | 
| Create folder | OK   |  OK   | 
| Copy | Bad Request   |  OK   | 
| Move | Not posible   |  OK   |
| Rename | OK   |  OK   |  
| Av. offline | OK   |  OK   | 
| Backgr (camera uploads) | (Conflict in upload, correct detected)   | OK | 
| Refresh account | OK   | OK | 

Here we have two levels of problem:

**High Severity**: In 9.0, the sharing option crashes. No options to do anything. We can justify as non-supported version but the crashes rate will increase. In both options, if the URL is https, app crashes after typing it (` java.lang.IllegalArgumentException: The host activity must implement com.owncloud.android.ui.dialog.SslUntrustedCertDialog.OnSslUntrustedCertListener`). 

**Medium/Low Severity**: Uploads are not performed (always set as `Conflict`). Move and Copy do not work either. This problems do not lead the app to an unstable/unconsistent status, but can bother users, mainly in case of uploads. 

The problems reported here do not happen in `master` branch with commit `ab17399` 

Summing up:

- Crash in sharing should be fixed as well as the https one. If sharing is not supported we can go for a disabling of the option or similar.
- Tests should be done with an 8.x server. Will check how/where to get it... Not sure how many users are stucked in such version, but after the experience of the endpoint switching, it worths a review. We should not forget it.
- As agreed, if the server version is lower than 10 (9.1 is also valid as limit), user will be warned of non-supported server version (this one to be done in a separate issue).

CC @michaelstingl @hosy @davigonz @abelgardep

**Comment by jesmrec** [comment]
As clear status of this:

- Clearing the (old) code of the app is a must in terms of re-architecting. Supporting old code makes the future-proof more difficult in terms of engineering.
- If we'd remove the support (as the new iOS app does) by avoiding old servers to connect to the app, this is a way of telling users: "You can not use the app you paid for". Dangerous, but this is also a business decision, not only a technical one.
- Technically, users of old servers must be disconnected after upgrading. Or spreading the world, warning them: "if your server is older than X, please do not upgrade to Android 2.14". A plan B is needed for such users that will upgrade even when their servers are not supported (redirecting to 2.13 in F-droid or similar).

These thoughts are based in our experience after switching to the new DAV endpoints the last year.  It will not be posible that everyone will be happy, whatever the final decision is. 

We need a clear agreement and a plan to roll it out.

Choices:

1. Remove all support oC < 10. Many users will be angry (and will ask for refunds). If users upgrade  "accidentally" the app, crashes will happen. Damaging us and our awesome market rates. 

2. Support all oC versions. Developers will be angry and the code less-qualitied. We must fix the bugs above, spending time to support versions that in fact, are EOL.

**Comment by abelgardep** [comment]
This will be merged here in https://github.com/owncloud/android/pull/2756. I think that we can close this now.
